src/app/services/player.service.ts

```diff
@@ -144,51 +172,112 @@ export class PlayerService {
     this.player.volume =
       event.offsetX / (event.target as HTMLElement).offsetWidth;
     this.musicVolume$.next(this.player.volume * 100);
+    this.isVolume = !!this.player.volume;
   }
 
-  putOnRepeat(): void {
-    this.isRepeat = !this.isRepeat;
-    this.player.loop = !this.player.loop;
+  checkMusicEnd(): void {
+    if (this.player.currentTime === this.player.duration) {
+      if (!this.player.loop) {
+        if (this.checkTrackExistence(this.currentTrackNumber + 1)) {
+          this.switchTrack(SwitchPlayerActionEnum.SWITCH_NEXT);
+        } else {
+          this.checkRepeat();
+        }
+      }
+    }
   }
 
-  checkMusicEnd(): void {
-    if (
-      this.player.currentTime === this.player.duration &&
-      !this.player.loop &&
-      this.checkTrackExistence(this.currentTrackNumber + 1)
-    ) {
-      this.switchTrack(SwitchPlayerActionEnum.SWITCH_NEXT);
-    } else if (
-      this.player.currentTime === this.player.duration &&
-      !this.player.loop
-    ) {
-      this.player.pause();
-      this.isPlay$.next(false);
-      this.stop$.next();
+  checkRepeat(): void {
+    if (!this.isRepeat) {
+      if (this.player.currentTime === this.player.duration) {
+        this.player.pause();
+        this.isPlay$.next(false);
+        this.stop$.next();
+      }
     }
+    if (this.isRepeat === 1) {
+      if (this.trackList$.getValue()) {
+        const newTrack = this.checkTrackExistence(0);
+        if (newTrack) {
+          this.currentTrackInfo$.next(newTrack);
+          this.switchPlayerAction();
+          this.currentTrackNumber = 0;
+        }
+      } else {
+        this.player.currentTime = 0;
+      }
+    }
+  }
+
+  putOnRepeat(): void {
+    this.isRepeat = this.repeatGen.next().value as number;
+    this.player.loop = this.isRepeat === 2;
   }
 
   closeAudioContext(): void {
     this.context.close();
+    this.stop$.next();
+    this.die$.next();
   }
 
   checkTrackExistence(
     index: number
   ): TrackById | NewAlbumTracksModel | null | undefined {
-    return this.trackList$.getValue()?.items[index]?.track;
+    if (this.isShuffle$.getValue()) {
```

Comment:
return this.isShuffle$.getValue()
    ? this.shuffleTrackList.items[index]?.track
    : this.trackList$.getValue()?.items[index]?.track;